HedvigInsurance/web-onboarding
# CPS / Get quote details data

## What?
Using the new generic `data` property on quotes instead of `quoteDetails`, and querying for exactly that where we need no more data than that.

- Types for the different `data` objects we expect to get on the different quotes, in `quoteDetailsDataTypes.ts`
- New `QuoteDetailsData` query, and generated hooks for that, to be able to make _one_ request to only get the `data`
- `useQuoteCartIdFromUrl` hook so that we don't have to do the `useParams<{ id: string }>()` everywhere
- `QuoteDetails` component remade to not get its data as props but fetch is directly in the component
- `getQuoteDetails.ts` refactored
- Unit tests for `getQuoteDetails` function. These are pretty suboptimal though, and it would make more sense IMO to at some point replace these with component tests with React Testing Library
- A few new text keys and translations for quote details labels and values
- Quote details Storybook story is modified to work with the new data fetching structure for `QuoteDetails` (this includes adding some mocked responses to the new `QuoteDetailsData` query)


## Why?

- The generic `data` object on `quote`s replaces the now deprecated `quoteDetails` property (reason is making the quotes generic on the back-end, an IPC project), which means we don't get any types from our schema telling us what quote details properties we can expect. Hence types being declared here.
- Using the `data` directly from the `QuoteDetails` component, with a query of its own, follows the [Apollo best practices](https://www.apollographql.com/docs/react/data/operation-best-practices/#query-only-the-data-you-need-where-you-need-it) that we've been trying to follow with the new checkout components 



**Ticket(s): [GRW-730]**





## Screenshots / recordings 

**Swedish House Quote:**
<img width="959" alt="image" src="https://user-images.githubusercontent.com/42962286/156167553-cba782ce-b390-4692-a403-aec1e4fe06e3.png">

**Norwegian Combo Quote:**
<img width="947" alt="image" src="https://user-images.githubusercontent.com/42962286/156167660-6706183a-e352-4524-8731-6b65c6ed6e0a.png">


**Danish Home Contents Student:**
<img width="916" alt="image" src="https://user-images.githubusercontent.com/42962286/156168257-954647d2-4b28-4412-a32a-2f47b4530654.png">




## How to test
Check out this branch and run Storybook to view the `QuoteDetails` component with mocked data.







[GRW-730]: https://hedvig.atlassian.net/browse/GRW-730?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
